x/exchange/fulfillment_test.go

```diff
@@ -0,0 +1,6220 @@
+			expPanic: "invalid indexed amount \"addr1\" for address \"-1neg\": cannot be zero or negative",
+		},
+		{
+			name: "one addr zero amount",
+			receiver: &IndexedAddrAmts{
+				addrs: []string{"addr1"},
+				amts:  []sdk.Coins{{{Denom: "zero", Amount: sdkmath.NewInt(0)}}},
+				indexes: map[string]int{
+					"addr1": 0,
+				},
+			},
+			expPanic: "invalid indexed amount \"addr1\" for address \"0zero\": cannot be zero or negative",
+		},
+		{
+			name: "one addr positive amount",
+			receiver: &IndexedAddrAmts{
+				addrs: []string{"addr1"},
+				amts:  []sdk.Coins{coins("1one")},
+				indexes: map[string]int{
+					"addr1": 0,
+				},
+			},
+			expected: []banktypes.Input{
+				{Address: "addr1", Coins: coins("1one")},
+			},
+		},
+		{
+			name: "two addrs",
+			receiver: &IndexedAddrAmts{
+				addrs: []string{"addr1", "addr2"},
+				amts:  []sdk.Coins{coins("1one"), coins("2two,3three")},
+				indexes: map[string]int{
+					"addr1": 0,
+					"addr2": 1,
+				},
+			},
+			expected: []banktypes.Input{
+				{Address: "addr1", Coins: coins("1one")},
+				{Address: "addr2", Coins: coins("2two,3three")},
+			},
+		},
+		{
+			name: "three addrs",
+			receiver: &IndexedAddrAmts{
+				addrs: []string{"addr1", "addr2", "addr3"},
+				amts:  []sdk.Coins{coins("1one"), coins("2two,3three"), coins("4four,5five,6six")},
+				indexes: map[string]int{
+					"addr1": 0,
+					"addr2": 1,
+					"addr3": 2,
+				},
+			},
+			expected: []banktypes.Input{
+				{Address: "addr1", Coins: coins("1one")},
+				{Address: "addr2", Coins: coins("2two,3three")},
+				{Address: "addr3", Coins: coins("4four,5five,6six")},
+			},
+		},
+		{
+			name: "three addrs, negative in third",
+			receiver: &IndexedAddrAmts{
+				addrs: []string{"addr1", "addr2", "addr3"},
+				amts: []sdk.Coins{
+					coins("1one"),
+					coins("2two,3three"),
+					{
+						{Denom: "acoin", Amount: sdkmath.NewInt(4)},
+						{Denom: "bcoin", Amount: sdkmath.NewInt(5)},
+						{Denom: "ncoin", Amount: sdkmath.NewInt(-6)},
+						{Denom: "zcoin", Amount: sdkmath.NewInt(7)},
+					},
+				},
+				indexes: map[string]int{
+					"addr1": 0,
+					"addr2": 1,
+					"addr3": 2,
+				},
+			},
+			expPanic: "invalid indexed amount \"addr3\" for address \"4acoin,5bcoin,-6ncoin,7zcoin\": cannot be zero or negative",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			orig := copyIndexedAddrAmts(tc.receiver)
+			var actual []banktypes.Input
+			testFunc := func() {
+				actual = tc.receiver.GetAsInputs()
+			}
+			assertions.RequirePanicEquals(t, testFunc, tc.expPanic, "GetAsInputs()")
+			assert.Equal(t, tc.expected, actual, "GetAsInputs() result")
+			if !assert.Equal(t, orig, tc.receiver, "receiver before and after GetAsInputs()") {
+				t.Logf("Before: %s", indexedAddrAmtsString(orig))
+				t.Logf(" After: %s", indexedAddrAmtsString(tc.receiver))
+			}
+		})
+	}
+}
+
+func TestIndexedAddrAmts_GetAsOutputs(t *testing.T) {
+	coins := func(coins string) sdk.Coins {
```

Comment:
Does this has to be redefined multiple times?

Reply:
These are defined individually in each unit test in a way that makes sense for the test.

E.g., this one takes in a string to convert, while another might take in a single amount and denom.

Reply:
These don't seem like a big impact on readability or maintainability to me ...